matplotlib/mpl_plotting/mpl_plotting.py

```diff
@@ -0,0 +1,110 @@
+If you want a more challenging exericse, try creating the plot without looking
+at the sub-steps.
+
+Bonus
+~~~~~
+
+Add some titles and axes to the plot.
+
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+import seaborn
+
+# Create some sample data
+x1 = np.linspace(0 ,10, 100)   # generate 100 numbers between 0 and 10
+y1 = np.sin(x1)                # element-wise sin of each element in x1
+e1 = np.random.ranf(100) * 5.0 # generate random floats in [0.0, 5.0)
+y1_noisy = y1 + e1             # make a noisy sin wave
+x2 = np.linspace(0, 10, 100)   # generate 100 numbers between 0 and 10
+y2 = np.sin(x2)                # element-wise sin of each element in x2
+e2 = np.random.ranf(100)*3.0   # generate noise with amplitude 3.0
+offset = 3.0                   # constant offset
+y2_noisy = y2 + offset + e2    # make a noisy sin wav offset from the first one
+
+# 1. Create a new figure.
+plt.figure()
+
+# 2. Create three columns and two rows of subplots and select the first one.
+plt.subplot(2,3,1)
+# 3. In the top, left subplot, plot y1_noisy using a line plot with black
+#    triangle markers.
+
+plt.plot(y1_noisy,'k-^')
+
+# 4. Overwrite the top left subplot with a plot of y1_noisy using a line plot
+#    with a red dashed line.
+plt.cla()
+plt.plot(y1_noisy,'r--')
+plt.title('Sine Wave')
+plt.ylabel('y1_noisy')
+# 5. In the top, middle subplot, plot y2_noisy using a line plot with
+#    green circle markers.
+plt.subplot(2,3,2)
+plt.plot(y2_noisy,'go')
+plt.title('Offset Sine')
+plt.ylabel('y2_noisy')
+# 6. In the top, right subplot, plot y1_noisy vs. y2_noisy using a scatter plot
+#    with blue circles.
+plt.subplot(2,3,3)
+plt.scatter(y1_noisy,y2_noisy,c='b')
+plt.title('Sine vs. Offset Sine')
+plt.xlabel('y1_noisy')
+plt.ylabel('y2_noisy')
+# 7. In the bottom, left subplot, plot the histogram of (y2_noisy - y1_noisy)
+plt.subplot(2,3,4)
+plt.hist(y2_noisy-y1_noisy)
+plt.title('Sine - Offset Sine')
+plt.ylabel('y2_noisy-y1_noisy')
+
+# 8. In the bottom, middle subplot, plot the histograms side by side for
+#    y1_noisy and y2_noisy (set hold to True again first).
+plt.subplot(2,3,5)
+plt.hist([y1_noisy,y2_noisy], label=['y1_noisy','y2_noisy'])
+plt.title('Sine & Offset Sine')
+plt.legend()
+# 9. In the bottom, right subplot, plot the correlation matrix of
+#        x1, y1, e1, y1_noisy, x2, y2, e2, y2_noisy
+#    as an image.
+#    Hint: syntax for correlation matrix is
+#        result = np.corrcoef([arr1, arr2, arr3, ...])
+plt.subplot(2,3,6)
+matrix = np.corrcoef([x1,y1,e1,y1_noisy,x2,y2,e2,y2_noisy] )
+labels=['x1','y1','e1','y1_noisy','x2','y2','e2','y2_noisy']
+x=[1,2,3,4,5,6,7]
+
+mask = np.zeros_like(matrix)
+
+w=seaborn.heatmap(matrix, cmap='viridis', vmax=1.0, vmin=-1.0 , mask = mask, linewidths=2.5)
+w.set_xticklabels(labels,rotation=45)
+w.set_yticklabels(labels[::-1],rotation=45)
+plt.imshow(matrix,interpolation='none')
+# 10. Add a colorbar to the subplot.
+plt.colorbar()
```

Comment:
The only thing I don't like in your solution is the awkward double colorbar.  
<img width="360" alt="screen shot 2018-02-09 at 4 10 21 pm" src="https://user-images.githubusercontent.com/852629/36052551-c3f0dbc4-0db3-11e8-8e8f-30d0e6c2ee82.png">
I think it can be eliminated by removing this line.  I'm pretty sure that `imshow` includes the colorbar by default, and maybe `matshow` doesn't.